skeleton/src/main/java/com/dropbox/kaiken/skeleton/core/KaikenUserServicesProvider.kt

```diff
@@ -3,59 +3,61 @@ package com.dropbox.kaiken.skeleton.core
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.scan
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
 
 /**
- * Skeleton implementation of [UserServicesProvider] that registers itself with a [UserManager].
+ * Skeleton implementation of [UserServicesProvider] that registers itself with a [UserStore].
  *
  */
+@OptIn(InternalCoroutinesApi::class, kotlinx.coroutines.ExperimentalCoroutinesApi::class)
 class KaikenUserServicesProvider
 constructor(
     private val applicationServices: AppServices,
     private val userServicesFactory: (AppServices, SkeletonUser) -> UserServices,
 ) : SkeletonUserServicesProvider {
-    private val userServicesMap = mutableMapOf<String, KaikenUserServices>()
+    private lateinit var userServices: Flow<Map<String, KaikenUserServices>>
+    private val mutex = Mutex(true)
 
     init {
         // TODO Mike figure out if we want another scope to launch from
         applicationServices.cast<CoroutineScopeBindings>().coroutineScopes().mainScope.launch {
-            (applicationServices as KaikenAppServices).userManager().getUserState()
-                .collect { userState ->
-                    userState.usersRemoved.forEach { user ->
-                        teardownUserServicesOf(user.userId)
+            userServices = (applicationServices as KaikenAppServices).userStore().getUserEvents()
+                .scan(emptyMap<String, KaikenUserServices>()) { prev, next ->
+                    val result = mutableMapOf<String, KaikenUserServices>()
+                    next.usersRemoved.forEach { user ->
+                        prev[user.userId]?.getUserTeardownHelper()?.teardown()
                     }
-                    // TODO Mike figure out what we want to do for the base user type
-                    userState.usersAdded.forEach { user ->
-                        initUserServicesOf(
-                            SkeletonUser(
-                                user.userId,
-                                SkeletonOauth2(user.accessToken),
-                                user.isActiveUser
-                            )
-                        )
+
+                    next.usersAdded.forEach { user ->
+                        val services = userServicesFactory(applicationServices, SkeletonUser(user.userId, user.accessToken)) as KaikenUserServices
+                        result[user.userId] = services
                     }
-                }
-        }
-    }
 
-    override fun initUserServicesOf(user: SkeletonUser) {
-        synchronized(userServicesMap) {
-            val userServices: KaikenUserServices =
-                userServicesFactory(applicationServices, user) as KaikenUserServices
-            userServicesMap[user.userId] = userServices
+                    result
+                }
+                .drop(1)
+            mutex.unlock()
```

Comment:
nit still think `CompletableDeferred` is the suspendy version of a mutex.

Reply:
Interesting, I'll give it a try!

Reply:
maybe it doesn't makes sense for you here but you can have the ability to await the completabledeferred inside of a suspending context.